### PR TITLE
Show users their own resources

### LINKED ISSUE #12: Show users their own published resources
## User story

**As a** user
**I want** to see the resources I suggested
**So that** I don't have to search through the whole list

## Acceptance criteria

- **Scenario**: user sees their published resources
  - **Given**: I am logged in as a user
  - **And**: I have had some suggestions published
  - **When**: I visit the account page
  - **Then**: I see the published resources I suggested

- **Scenario**: user does not see their draft resources
  - **Given**: I am logged in as a user
  - **And**: I have had some suggestions that have no yet been published
  - **When**: I visit the account page
  - **Then**: I do not see the draft resources I suggested

### PR DESCRIPTION
**Pull request by Emmanuel-Dominic**
This is a:

- [x] ✨ **New feature** - new behaviour has been implemented
- [ ] 🐛 **Bug fix** - existing behaviour has been made to behave
- [ ] ♻️ **Refactor** - the behaviour has not changed, just the implementation
- [ ] ✅ **Test backfill** - tests for existing behaviour were added but the behaviour itself hasn't changed
- [ ] ⚙️ **Chore** - maintenance task, behaviour and implementation haven't changed

### Description
- [x] When you log into the application while you have published resources, navigate to the accounts page to view your personal resources, while as an administrator, you will see your personal published and draft resources.
- [x] Users are able to view their published resources when logged in as normal authenticated users while admins can also see draft resources.

- **Purpose** - 
- [x] Users can see the resources they suggested without having to search through the entire list of resources in the application.

- **How to check** - 
- [x] Login into the application.
- [x] Write some suggestions to be published yet if you have logged in as an admin
- [x] If you're an admin, you will be able to see a list of your published resources as well as your draft resources on the account page.
- [x] If you're a normal authenticated user, you will see only the published resources.
### Links
- [x] resolve #12 

### Author checklist

- [x] I have written a title that reflects the relevant ticket
- [x] I have written a description that says what the PR does and how to validate it
- [x] I have linked to the project board ticket (and any related PRs/issues) in the Links section
- [x] I have added a link to this PR to the ticket
- [x] I have made the PR to `main` from a branch named `feature/show-users-resource`
- [x] I have manually tested that the app still works correctly
- [x] I have requested reviewers here and in my team chat channel

- [ ] I have spoken with my PM or TL about any parts of this task that may have become out-of-scope, or any additional improvements that I now realise may benefit my project
- [x] I have added tests, or new tests were not required
- [x] I have updated any documentation (e.g. diagrams, schemas), or documentation updates were not required

### CODE REVIEW

File: client/src/services/resourceService.test.js [hunk 1]
```diff
@@ -11,13 +11,13 @@ describe("ResourceService", () => {
...
-				rest.get("/api/resources", (req, res, ctx) => {
-					request = req;
+				rest.get("/api/resources", async (req, res, ctx) => {
+					request = await req;
 					return res(ctx.json({ resources: [] }));
 				})
 			);
 			await service.getDrafts();
-			expect(request.url.searchParams.get("draft")).toBe("true");
+			await expect(request.url.searchParams.get("draft")).toBe("true");
```
Comment (textbook) [inline client/src/services/resourceService.test.js:20]: ❓ Not sure why these changes have been made. Neither `req` nor `expect(...).toBe(...)` is a promise. Did this test fail or timeout at some point?

File: client/src/services/resourceService.test.js [hunk 2]
```diff
@@ -129,4 +129,42 @@ describe("ResourceService", () => {
...
+			await server.use(
+				rest.get("/api/resources", async (req, res, ctx) => {
+					return res(ctx.json({ resources: [] }));
+				})
+			);
+			await expect(
+				service.getAccountResources({ account: true, page: 1, perPage: 5 })
+			).resolves.toEqual({ resources: [] });
+		});
+	});
```
Comment (textbook) [inline client/src/services/resourceService.test.js:169]: ❌ this exact same test appears twice with different names (and slightly different arrow function formatting) - and in neither case does the response actually include any resources.

File: e2e/integration/account.test.js [hunk 3]
```diff
@@ -5,6 +5,7 @@ it("shows the user's details", () => {
 	cy.findByRole("link", { name: /account/i }).click();
 	cy.findByText(email).should("exist");
 	cy.findByText("Cyf Admin").should("exist");
+	cy.findByRole("heading", { level: 1 }).should("contain.text", "Resources");
```
Comment (textbook) [inline e2e/integration/account.test.js:8]: ❓ This change seems unrelated to these changes

❌ There's no E2E test that _is_ related to these changes

File: server/docs/resources.yml [hunk 4]
```diff
@@ -7,6 +7,14 @@
       - authCookie: []
       - sudoToken: []
     parameters:
+      - in: query
+        name: account
+        description: |
+          Filter resources by authenticated user status.
```
Comment (textbook) [inline server/docs/resources.yml:13]: ❓ I don't think this _does_ filter by status. It filters by the source (hence my API suggestions above ☝️).

File: server/docs/resources.yml [hunk 5]
```diff
@@ -7,6 +7,14 @@
...
+
+          **Note**: Published resources, belonging to the authenticated user, are available to him/her as drafts are only accessible to admin users.
```
Comment (textbook) [inline server/docs/resources.yml:15]: 💡 I found this quite confusing, I think what you're trying to say is something like:
```suggestion
          **Note**: Published resources belonging to the authenticated user are available to all users; drafts are only accessible to admin users.
```

File: server/resources/resources.test.js [hunk 6]
```diff
@@ -242,7 +242,7 @@ describe("/api/resources", () => {
 
 			await anonAgent
 				.get("/api/resources")
-				.query({ draft: true })
+				.query({ draft: true, account: false })
```
Comment (textbook) [inline server/resources/resources.test.js:245]: ❓ Isn't this the implicit default?

❌ There are no tests for the new behaviour

File: server/resources/resourcesController.js [hunk 7]
```diff
@@ -27,11 +27,17 @@ router
 			}).unknown(),
 		}),
 		asyncHandler(async (req, res) => {
-			const { draft, page, perPage } = req.query;
-			if (draft && !req.superuser && !req.user?.is_admin) {
+			const { draft, account, page, perPage } = req.query;
```
Comment (textbook) [inline server/resources/resourcesController.js:30]: 💡 I'd go with:
```suggestion
			const { account, draft, page, perPage } = req.query;
```

File: server/resources/resourcesRepository.js [hunk 8]
```diff
@@ -49,8 +69,27 @@ export const count = async ({ draft }) => {
...
+	offset,
+}) => {
+	let query;
+	let param = [limit, offset];
+	if (is_admin && account) {
+		query = pagedResourceAdminQuery;
+		param.unshift(id);
+	} else if (!is_admin && account) {
+		query = pagedResourceUserQuery;
+		param.unshift(id);
```
Comment (textbook) [inline server/resources/resourcesRepository.js:87]: 💡 The first two cases are basically the same. This would be clearer in my opinion if `account` was _first_, but another way to do it is collapse them entirely:
```suggestion
	if (account) {
		query = is_admin ? pagedResourceAdminQuery : pagedResourceUserQuery;
		param.unshift(id);
```

**Review by textbook — CHANGES_REQUESTED** [review]
## UX/UI

✅ At a high level, this functionality works. As an admin I see:

Screenshot 2024-01-25 at 09 34 42

and as a non-admin I see:

Screenshot 2024-01-25 at 09 36 09

❓ Should there be a heading for this section?

❓ Could we show something different to the paginated table if they have no resources/suggestions (or skip this section entirely)?

❓ I wonder if it would be helpful to indicate that a resource is a draft more explicitly than just showing the publish button? The other place we show draft resources, they're explicitly _on the drafts page_ (and remember admins would stop seeing drafts here with #11, but non-admins would _start_ seeing drafts with #15).

## API

❓ I've made a few suggests around the REST API, but actually overall I wonder if adding more functionality to `GET /api/resources` is the right approach. What if, instead, this was `GET /api/users/<userid>`, which returned information about the user _including their resources_? That might better align with the usage of the Account page. You'd still have the implementation in the `resourcesService`, but then this would be consumed by the `usersService` rather than the `resourcesController`.

## Implementation

More details comments inline with the code, but at a high level:

✅ you've updated all of the correct parts of the application, including the docs, and put the functionality in sensible places - this is great!

❌ insufficient testing - I'd guess you mostly just wrote the implementation, then applied the minimal fixes to any tests that broke, rather than trying to test-_drive_ the new functionality. At the very least we need an E2E test to cover this new functionality.

---

**Key**:

- ✅ approval, nothing needs doing before merge
- ❌ problem, needs fixing before merge
- ❓ question, needs answering before merge
- 💡 suggestion, can be followed or ignored

**Comment by Emmanuel-Dominic** [comment]
Hello @textbook, Thanks for your feedback, however, I would like to clarify this statement for clarity `I wonder if it would be helpful to indicate that a resource is a draft more explicitly than just showing the publish button? The other place we show draft resources is they're explicitly on the drafts page (and remember admins would stop seeing drafts here with issue #11, but non-admins would start seeing drafts with issue #15)`.

So, is my picked up issue #12 intended to cover issue #11 while leaving room for the implementation of issue #15 or should it be treated as a standalone issue leaving room for both issues #11 and #15 to be addressed?

**Comment by textbook** [comment]
Specifically within this story as proposed, per the pictures I shared above: it's _implicit_ to the admin users that some of the resources they're seeing are drafts, because they have a publish button. So I wonder, right now, if that should me made explicit or whether we can rely on them to figure it out (or immediately publish their own resources so never see drafts here, or simply not care, etc.). I mention the other two stories because we should always think about our work _in context_, to help make the best decisions - if we delivered #11 (so admins don't see drafts) but never #15 (so non-admins don't see them either) then we'd never see drafts in this view at all, for example.

**Review by textbook — CHANGES_REQUESTED** [review]
Thank you for iterating based on the feedback provided.

❌ The one thing that needs to be fixed before we can consider the task complete is: the pipeline is failing, because the low-level tests don't pass (and if they did, the high-level test would also fail). _Use the tests_ to help you develop, at the very least you should be running them before you push (use `npm run ship` to run all of the automated checks). **Please get these passing.**

---

One other thing to consider for future work:

💡 When I suggested making the request `GET /users/:id`, I had intended that the return value would be _a representation of the user_, not _just_ an array of resources. As a minimal response for our needs right now, `{ "resources": [...] }`. But also the purpose of splitting the backend into separate directories is to represent the _domains_ in our codebase, `usersRepository` shouldn't be doing queries that are primarily on the `resources` collection. For example, the user service could look something like this:

```javascript
export async function getByIdWithResources(id, pagination) {
  const user = await getById(id);
  const resources = await resourceService.getAll({ source: id }, pagination);
  return { resources, ...user };
}
```
although as I'm writing that I'm thinking:
- maybe `GET /user/:id?page=2&perPage=15` doesn't really make a lot of sense, we're paginating the sub-resource - could we assume that an individual user won't have supplied so many resources that pagination would be needed, or we make it that `GET /user/:id` only ever includes their last X resources and you have to hit some other endpoint if you want to see more?
- what should the permissions for this endpoint be? I should see drafts if it's my own user, or I'm an admin, but I'd say a non-admin or unauthenticated user (already prevented by `authOnly`) shouldn't be able to see another user's drafts.

API design is fun!

**Comment by Emmanuel-Dominic** [comment]
Hello @textbook, thank you for providing this insight 
--------------------------------------------

💡 As I see it, it makes sense, since I buy the idea of returning the user data with the loadingMore option, while displaying only the most recent resources without passing the pagination parameter to the URL.
- Let me work on that shortly.

**Comment by textbook** [comment]
No problem! If you'd like to work on this further that's great, it's all good practice, but for the Tech Products interview process just focus on getting the pipeline to ✅

File: client/src/services/userService.test.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
+import { resourceStub } from "../../setupTests";
+
+import UserService from "./userService";
+
+describe("UserService", () => {
```
Comment (textbook) [inline client/src/services/userService.test.js:5]: Fun fact: if you comment out or delete the whole implementation of the actual `getByUser` method, these tests still pass. You're replacing the real implementation with a test double, the mock function created with `jest.fn()`, then calling _the test double_ and asserting on its behaviour (i.e. _"does the test double do what I literally just asked it to do?"_) This is why if you look at the test coverage report a big chunk of the service simply isn't covered by the tests.

Comment (textbook) [reply client/src/services/userService.test.js:7]: FYI this continues to have the problem that you're testing a test double; you've replaced the entire real implementation with `jest.mock("./userService");`. It's absolutely crucial when writing tests to ensure that your actual implementation is involved - even if you write the tests afterwards for code that's (or that you think is) working, temporarily break the implementation _on purpose_ to watch them fail. Test doubles are for _collaborators_, things the system under test uses, not to entirely replace the thing you're supposed to be testing.
